### DISCUSSION TITLE
array is listed on every details page but should only be listed on relaited details page

### DISCUSSION THREAD

**Discussion by marcterre** [neuefische/questions#467] [Web] [answered]
Hello everyone, please have a look at my [preview](https://capstone-project-by1nvjubn-marcterre.vercel.app/)
You are able to create projects which will be listed on the landingpage. In a project you are able to add a new view which will be listed on the project details page. The problem is the view will also be listet in every other project. If you look at my code its obvious.. i create two arrays with the data and map over them. I think i have to filter something and map over the filtered one... but i dont know what to do exactly 😵‍💫

Thats how i get the data from handle submit and pass them in a array:

```
  const [projects, setProjects] = useLocalStorageState("projects", {defaultValue: [], });
  const [views, setViews] = useLocalStorageState("views", { defaultValue: [] });


function addNewProject(newProject) {
    setProjects((oldProjects) => [
      {
        ...newProject,
        id: uuidv4(),
        slug: newProject.name,
      },
      ...oldProjects,
    ]);
  }

  function addNewView(newView) {
    setViews((oldViews) => [
      {
        ...newView,
        id: uuidv4(),
        viewSlug: newView.name,
      },
      ...oldViews,
    ]);
  }
```

Thats how i map over the projects. They will be listet on the landingpage:
```
export default function ProjectItem({ projects }) {
  return (
    <List>
      {projects.map((project) => (
        <ListItem key={uuidv4()}>
          <Link
            href={`/project-details/${project.slug}`}
            onClick={() => console.log(project.id)}
          >
            <div>
              <StyledToolsIcon />
              {project.name}
            </div>
          </Link>
        </ListItem>
      ))}
    </List>
  );
}
```

Thats how i map over my views . They will be listed on every project details page:
```
 export default function ViewItem({ views, currentProject }) {

  return (
    <List>
      {views.map(({ id, name, viewSlug }) => (
        <ListItem key={id}>
          <Link
            href={`/project-details/${currentProject.slug}/view-details/${viewSlug}`}
          >
            <div>{name}</div>
          </Link>
        </ListItem>
      ))}
    </List>
  );
}
```
Hope thats enough.
feel free to contact me directly

**Comment by mbosselmann** [comment]
Hi Marc,
currently you store both your projects and your views in the local storage under two keys. These have no reference to each other and therefore it is not clear in the detail view which views should be displayed. Your solution can be done this way, but then there should be a reference at each view to which project the view refers to. You can set this reference by adding the `projectId` of the current project in your `addNewView` function for a new view.

```js

  function addNewView(newView) {
    setViews((oldViews) => [
      {
        ...newView,
        id: uuidv4(),
        viewSlug: newView.name,
        projectId: currentProject.id,
      },
      ...oldViews,
    ]);
  }
```

In your details page you can then integrate the filter functioniality:

```js
          <ViewItem
            views={views.filter((view) => view.projectId === currentProject.id)}
            projects={projects}
            currentProject={currentProject}
          />
```

I hope this does help to fix your problem. 

When I looked through your code to understand your process, I noticed that you set as slug the name of the project. Instead of giving this extra key to you project object, you can also use the id and this would work as well. Because by using the project name you get app paths that can have spaces, something we want to avoid when creating paths for our app.

And you could also use the details page to search for the current project, because only this page needs this information currently or? :)

**Reply by marcterre** [reply]
Hey Mareike appreciate your effort!
Now if want to add a new view it throws an error "currentProject is undefined". 
I dont know why -> [PR](https://github.com/marcterre/capstone-project/pull/13)

**Comment by shebtastic** [comment, marked as answer]
Currently the function `addNewView` does not add any key that creates a reference between a view and a project.
You have different options here:

1. Stick with split lists of views and objects, but add something like `projectId` to your view datastructure:
   ```js
   {
     "name": "Bla",
     "description": "",
     "sketch": "/Example.png",
     "projectId": "713876d0-41cd-4fa4-bb02-2ea3b714296c",
     "viewId": "23b4c9ef-f982-4a38-9618-3de2c02351a6"
   }
   ```
   This would also mean that when you open a project and want to display only the views that are linked to this project, you would have to filter the views that have the fitting `projectId`.
2. Merge the states to a single state, where the `views` array is part of your projects array. This would be a bigger change.
   ```js
   {
     "name": "Test",
     "description": "",
     "sketch": "/Example.png",
     "views": [],
     "id": "713876d0-41cd-4fa4-bb02-2ea3b714296c"
   }
   ```

Any of these changes would require you to ideally move the `pages/project-details/add-new-view/index.js` to `pages/project-details/[id]/add-new-view/index.js`. That way both of the pages have access to the path parameter you can get from the router.

**Reply by marcterre** [reply]
I choosed the first method and now it works. I think the essential was to move the  add-new-view in project-detetails/[id]/  Didnt know it was that important 🥲 and then of course adding the projectId. This way currentProject.id is also not undefined anymore.
Thank you a lot!

**Comment by kisa5** [comment]
Hi i have pretty much the same problem as Marc. I can add projects on my landing page and when I go to the project page I can add related tasks. The error is that every time I add a task it is added to all project pages.

Passing my data:
```
const [project, setProject] = useLocalStorageState("project", {
    defaultValue: [],
  });

  function addProject(projectname) {
    setProject([projectname, ...project]);
  }

  const [task, setTask] = useLocalStorageState("task", {
    defaultValue: [],
  });
  function addTask(taskname) {
    setTask([taskname, ...task]);
  }
```
mapping my projects:
```
export default function ProjectCard({ project }) {
  const router = useRouter();
  return (
    <ProjectList>
      {project.map((project, index) => (
        <ProjectItem key={index}>
          <Link href={`/project/${project.slug}`}>{project.title}</Link>
        </ProjectItem>
      ))}
    </ProjectList>
  );
}
```
mapping my tasks:
```
export default function Task({ addTask, task }) {
  const router = useRouter();
  const { slug } = router.query;

  return (
    <>
      <BackButton>
        <Link href="/"> ⬅️ </Link>
      </BackButton>
      <StyledList>
        <Title>{slug}</Title>
        {task.map((task, index) => (
          <TaskItem key={index}>{task.task}</TaskItem>
        ))}
      </StyledList>
      <AddTask addTask={addTask} />
    </>
  );
}
```
thanks in advance!

**Comment by F-Kirchhoff** [comment]
It looks like you have only one "task" state in your localstorage. That is why you cannot separate tasks from another in projects. Your projects don't know which task belongs to them, there is only one list of tasks independent from your projects. You need to add each tasks to a specific project, such that this project alone has the owernship over that task. Your Projects need to look something like this:
```js
{
  id: 1337,
  name: "test",
  tasks: [task1,task2,task3]
}
```
Then you need to update the Project if you add a task to that specific task list.
